reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java

```diff
@@ -515,16 +531,20 @@ boolean checkTerminated(boolean d, boolean empty) {
 			if (d) {
 				Throwable e = error;
 				if (e != null && e != Exceptions.TERMINATED) {
-					CONNECTION.compareAndSet(parent, this, null);
-					e = Exceptions.terminate(ERROR, this);
+					if (parent.resetUponSourceTermination) {
+						CONNECTION.compareAndSet(parent, this, null);
+						e = Exceptions.terminate(ERROR, this);
```

Comment:
Not completely certain about this one - it might require special handling around concurrency when `drain()` is called in parallel and `parent.resetUponSourceTermination` is false. It's required to preserve the actual error to be able to replay the error to late subscribers, but it's eliminating the guard behaviour of single delivery.